File: OpenDentBusiness/TableTypes/EmailMessage.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace OpenDentBusiness {
	///<summary>An outgoing email message is stored here.</summary>
	[Serializable]
	public class EmailMessage:TableBase {
		///<summary>Primary key.</summary>
		[CrudColumn(IsPriKey=true)]
		public long EmailMessageNum;
		///<summary>FK to patient.PatNum</summary>
		public long PatNum;
		///<summary>Single valid email address. Bcc field might be added later, although it won't be very useful.  We will never allow visible cc for privacy reasons.</summary>
		public string ToAddress;
		///<summary>Valid email address.</summary>
		public string FromAddress;
		///<summary>Subject line.</summary>
		public string Subject;
		///<summary>Body of the email</summary>
		[CrudColumn(SpecialType=CrudSpecialColType.TextIsClob)]
		public string BodyText;
		///<summary>Date and time the message was sent. Automated at the UI level.</summary>
		[CrudColumn(SpecialType=CrudSpecialColType.DateT)]
		public DateTime MsgDateTime;
		///<summary>0=neither, 1=sent, 2=received.</summary>
		public EmailSentOrReceived SentOrReceived;
		///<summary>Not a database column.</summary>
		[CrudColumn(IsNotDbColumn=true)]
		public List<EmailAttach> Attachments;

		///<summary>Constructor</summary>
		public EmailMessage(){
			Attachments=new List<EmailAttach>();
		}

		public EmailMessage Copy() {
			EmailMessage e=(EmailMessage)this.MemberwiseClone();
			e.Attachments=new List<EmailAttach>();
			for(int i=0;i<Attachments.Count;i++) {
				e.Attachments.Add(Attachments[i].Copy());
			}
			return e;
		}
	}

	///<summary>0=Neither, 1=Sent, 2=Receivedceived. 3=Read, 4=WebMailReceived, 5=WebMailRecdRead, 6=WebMailSent, 7=WebMailSentRead</summary>
	public enum EmailSentOrReceived {
		///<summary>0 Unsent</summary>
		Neither,
		///<summary>1 For regular email only.</summary>
		Sent,
		///<summary>2 For regular email only.  Shows in Inbox.  Once it's attached to a patient it will also show in Chart module.</summary>
		Received,
		///<summary>3 For received regular email only.  Has been read.  Shows in Inbox.  Once it's attached to a patient it will also show in Chart module.</summary>
		Read,
		///<summary>4 WebMail received from patient portal.  Shows in OD Inbox and in pt Chart module.  Also shows in PP as a sent and unread WebMail msg.</summary>
		WebMailReceived,
		///<summary>5 WebMail received from patient portal that has been marked read.  Shows in the OD Inbox and in pt Chart module.  Also shows in PP as a sent and read WebMail.</summary>
		WebMailRecdRead,
		///<summary>6 Webmail sent from provider to patient.  Shows in Chart module and also shows in PP as a received and unread WebMail msg.</summary>
		WebMailSent,
		///<summary>7 Webmail sent from provider to patient and read by patient.  Shows in Chart module and also shows in PP as a received and read WebMail msg.</summary>
		WebMailSentRead,
		///<summary>8 Sent and encrypted using Direct. Required for counting messages in EHR modules g.1 and g.2, Automated Measure Calculation.</summary>
		SentDirect,
		///<summary>9 Received email matches application/pkcs7-mime mime type, but could not be decrypted.  Shows in Inbox.  The user can decrypt from FormEmailMessageEdit.  If the user has the correct private key, then the status will change to Read.</summary>
		ReceivedEncrypted,
		///<summary>10 Received email matches application/pkcs7-mime mime type and has been decrypted.  Shows in Inbox.  Once it's attached to a patient it will also show in Chart module.  When viewing inside of FormEmailMessageEdit, the XML body of the message shows as xhtml instead of raw.  Still need to work on supporting collapsing and expanding, as required for meaningful use in 2014.</summary>
		ReceivedDirect,
		///<summary>11 For received direct messages.  Has been read.  Shows in Inbox.  Once it's attached to a patient it will also show in Chart module.  When viewing inside of FormEmailMessageEdit, the XML body of the message shows as xhtml.</summary>
		ReadDirect
	}


}

File: OpenDentalWeb/OpenDental/src/com/opendental/odweb/client/tabletypes/EmailMessage.java
package com.opendental.odweb.client.tabletypes;

import com.google.gwt.xml.client.Document;
import com.opendental.odweb.client.remoting.Serializing;
import com.google.gwt.i18n.client.DateTimeFormat;
import java.util.Date;
import java.util.ArrayList;

//DO NOT MAKE CHANGES TO THIS FILE.  THEY WILL GET OVERWRITTEN BY THE CRUD.
public class EmailMessage {
		/** Primary key. */
		public int EmailMessageNum;
		/** FK to patient.PatNum */
		public int PatNum;
		/** Single valid email address. Bcc field might be added later, although it won't be very useful.  We will never allow visible cc for privacy reasons. */
		public String ToAddress;
		/** Valid email address. */
		public String FromAddress;
		/** Subject line. */
		public String Subject;
		/** Body of the email */
		public String BodyText;
		/** Date and time the message was sent. Automated at the UI level. */
		public Date MsgDateTime;
		/** 0=neither, 1=sent, 2=received. */
		public CommSentOrReceived SentOrReceived;
		/** Not a database column. */
		public ArrayList<EmailAttach> Attachments;

		/** Deep copy of object. */
		public EmailMessage deepCopy() {
			EmailMessage emailmessage=new EmailMessage();
			emailmessage.EmailMessageNum=this.EmailMessageNum;
			emailmessage.PatNum=this.PatNum;
			emailmessage.ToAddress=this.ToAddress;
			emailmessage.FromAddress=this.FromAddress;
			emailmessage.Subject=this.Subject;
			emailmessage.BodyText=this.BodyText;
			emailmessage.MsgDateTime=this.MsgDateTime;
			emailmessage.SentOrReceived=this.SentOrReceived;
			emailmessage.Attachments=this.Attachments;
			return emailmessage;
		}

		/** Serialize the object into XML. */
		public String serialize() {
			StringBuilder sb=new StringBuilder();
			sb.append("<EmailMessage>");
			sb.append("<EmailMessageNum>").append(EmailMessageNum).append("</EmailMessageNum>");
			sb.append("<PatNum>").append(PatNum).append("</PatNum>");
			sb.append("<ToAddress>").append(Serializing.escapeForXml(ToAddress)).append("</ToAddress>");
			sb.append("<FromAddress>").append(Serializing.escapeForXml(FromAddress)).append("</FromAddress>");
			sb.append("<Subject>").append(Serializing.escapeForXml(Subject)).append("</Subject>");
			sb.append("<BodyText>").append(Serializing.escapeForXml(BodyText)).append("</BodyText>");
			sb.append("<MsgDateTime>").append(DateTimeFormat.getFormat("yyyyMMddHHmmss").format(MsgDateTime)).append("</MsgDateTime>");
			sb.append("<SentOrReceived>").append(SentOrReceived.ordinal()).append("</SentOrReceived>");
			sb.append("</EmailMessage>");
			return sb.toString();
		}

		/** Sets all the variables on this object based on the values in the XML document.  Variables that are not in the XML document will be null or their default values.
		 * @param doc A parsed XML document.  Must be valid XML.  Does not need to contain a node for every variable on this object.
		 * @throws Exception DeserializeFromXml is entirely encased in a try catch and will throw exceptions if anything goes wrong. */
		public void deserialize(Document doc) throws Exception {
			try {
				if(Serializing.getXmlNodeValue(doc,"EmailMessageNum")!=null) {
					EmailMessageNum=Integer.valueOf(Serializing.getXmlNodeValue(doc,"EmailMessageNum"));
				}
				if(Serializing.getXmlNodeValue(doc,"PatNum")!=null) {
					PatNum=Integer.valueOf(Serializing.getXmlNodeValue(doc,"PatNum"));
				}
				if(Serializing.getXmlNodeValue(doc,"ToAddress")!=null) {
					ToAddress=Serializing.getXmlNodeValue(doc,"ToAddress");
				}
				if(Serializing.getXmlNodeValue(doc,"FromAddress")!=null) {
					FromAddress=Serializing.getXmlNodeValue(doc,"FromAddress");
				}
				if(Serializing.getXmlNodeValue(doc,"Subject")!=null) {
					Subject=Serializing.getXmlNodeValue(doc,"Subject");
				}
				if(Serializing.getXmlNodeValue(doc,"BodyText")!=null) {
					BodyText=Serializing.getXmlNodeValue(doc,"BodyText");
				}
				if(Serializing.getXmlNodeValue(doc,"MsgDateTime")!=null) {
					MsgDateTime=DateTimeFormat.getFormat("yyyyMMddHHmmss").parseStrict(Serializing.getXmlNodeValue(doc,"MsgDateTime"));
				}
				if(Serializing.getXmlNodeValue(doc,"SentOrReceived")!=null) {
					SentOrReceived=CommSentOrReceived.valueOf(Serializing.getXmlNodeValue(doc,"SentOrReceived"));
				}
			}
			catch(Exception e) {
				throw new Exception("Error deserializing EmailMessage: "+e.getMessage());
			}
		}

		/** 0=neither, 1=sent, 2=received. */
		public enum CommSentOrReceived {
			/** 0 */
			Neither,
			/** 1 */
			Sent,
			/** 2 */
			Received
		}


}
